parlai/core/torch_agent.py

```diff
@@ -1551,47 +1520,59 @@ def match_batch(self, batch_reply, valid_inds, output=None):
                 batch_reply[i]['text_candidates'] = cands
         return batch_reply
 
-    def last_reply(self, use_reply='label'):
+    def self_observe(self, self_message: Message) -> None:
         """
-        Retrieve the last reply from the model.
-
-        If available, we use the true label instead of the model's prediction.
+        Observe one's own utterance.
 
-        By default, batch_act stores the batch of replies and this method
-        will extract the reply of the current instance from the batch.
+        This is used so that the agent can incorporate its own response into
+        the dialogue history after a batch_act. Failure to implement this will
+        result in an agent that cannot hear itself speak
 
-        :param use_label:
-            default true, use the label when available instead of the model's
-            generated response.
+        :param self_message:
+            The message corresponding to the output from batch_act.
         """
-        # if the last observation was the end of an episode,
-        # then we shouldn't use it as history
-        if (
-            use_reply == 'none'
-            or not self.observation
-            or self.observation['episode_done']
-        ):
-            return None
+        use_reply = self.opt.get('use_reply', 'label')
+
+        if self.last_observation is None:
+            raise RuntimeError(
+                "You're self_observing without having observed something. Check if "
```

Comment:
idk that we should always throw a RuntimeError here? what if in interactive mode we want the model to talk first? can you double check that this doesn't break?

Reply:
We never have supported that. Going first is always a `__silence__` input, or just the personas, or whatever.

Yes, I'll need to check interactive.py.

Reply:
i can imagine this coming up in e.g. an mturk task where we want the model to come first, but i'm fine with just observing `__silence__` at that point